thalassa/api.py

```diff
@@ -65,24 +65,102 @@ def get_elevation_dmap(trimesh: gv.TriMesh, show_grid: bool = False) -> hv.Overl
-    xys=((3.0,37),(30,33)) #todo: read fixed stations from inputs
-    hs=hv.HoloMap({'P_{}'.format(i+1):get_timeseries_xy(*xy) for i,xy in enumerate(xys)})
-
-    #dynamic
-    hd=hv.DynamicMap(get_timeseries_xy,streams=[PointerXY(x=0,y=0,source=source)])
-    hd=hd.opts(color='k',line_width=2,line_dash='dotted')
-
-    #points
-    hp=hv.HoloMap({'P_{}'.format(i+1):hv.Points(xy).opts(color='r') for i,xy in enumerate(xys)})
-
-    return hd*hs.overlay(),hp.overlay()
+#----------------------------------------------------------------------------------------
+#time series
+#----------------------------------------------------------------------------------------
+#define a class to store data related to time series points
+class timeseries_data:
+      def __init__(self):
+         self.init=False
+      def clear(self):
+          self.init=False
+
+#function for extracting time series@(x,y) from data
+def extract_timeseries(x,y,sx,sy,data):
+    dist=abs(sx+1j*sy-x-1j*y)
+    mdist=dist.min()
+    nid=np.nonzero(dist==mdist)[0][0]
+    mdata=data['elev'].data[:,nid].copy()
+    return mdist,mdata
+
+#function to dynamically add or remove pts by double clicking on the map
+def add_remove_pts(x,y,data,dataset,fmt):
+    if fmt=='add pts':
+       if len(data.xys)==0:
+          mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+          hcurve=hv.Curve((data.time,mdata),'time','elevation')
+          if mdist<=data.mdist:
+             data.xys.append((x,y))
+             data.elev.append(mdata)
+             data.curve.append(hcurve)
+       else:
+          if data.xys[-1][0]!=x and data.xys[-1][1]!=y:
+             mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+             hcurve=hv.Curve((data.time,mdata),'time','elevation')
+             if mdist<=data.mdist:
+                data.xys.append((x,y))
+                data.elev.append(mdata)
+                data.curve.append(hcurve)
+    elif fmt=='remove pts':
+       if len(data.xys)>0:
+          xys=np.array(data.xys)
+          dist=abs(xys[:,0]+1j*xys[:,1]-x-1j*y)
+          mdist=dist.min()
+          if mdist<=data.mdist:
+             nid=np.nonzero(dist==mdist)[0][0]
+             data.xys=[k for i,k in enumerate(data.xys) if i!=nid]
+             data.elev=[k for i,k in enumerate(data.elev) if i!=nid]
+             data.curve=[k for i,k in enumerate(data.curve) if i!=nid]
+    else:
+       pass
+
+#get time series plots
+def get_timeseries(source,data,dataset,ymin,ymax,fmt):
+    #initialize timeseries_data
+    if data.init is False:
+       #find the maximum side length
+       x,y=dataset['SCHISM_hgrid_node_x'].data,dataset['SCHISM_hgrid_node_y'].data
+       e1,e2,e3=dataset['SCHISM_hgrid_face_nodes'].data.T
+       s1=abs((x[e1]-x[e2])+1j*(y[e1]-y[e2])).max()
+       s2=abs((x[e2]-x[e3])+1j*(y[e2]-y[e3])).max()
+       s3=abs((x[e3]-x[e1])+1j*(y[e3]-y[e1])).max()
+
+       #save data
+       data.sx, data.sy, data.x0, data.y0  = x, y, x.mean(), y.mean()
```

Comment:
1. Please one assignment per line
2. Perhaps we should consider renaming `x` -> `sx` and `y` -> `sy` (or vice versa) just to be more consistent

Reply:
1. See my comment above 
2. OK, I will change (x,y) -> (sx,sy) here.

Reply:
I check the code, and think (x, y) are still better names here because they are local temporary names. Global names should be used carefully.